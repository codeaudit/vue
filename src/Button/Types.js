export const standard = 'standard';
export const basic = 'basic';
export const important = 'important';
export const alt = 'alt';
export const basicAlt = 'basic-alt';
export const special = 'special';
