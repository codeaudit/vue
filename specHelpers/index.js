import React from 'react';
import ThemeProvider, {getTheme} from './../src/Theme';

export const applyTheme = (ComponentUnderTest, theme = getTheme()) => <ThemeProvider theme={theme}>{ComponentUnderTest}</ThemeProvider>;
