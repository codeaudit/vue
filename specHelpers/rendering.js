import toJson from 'enzyme-to-json';
import {mount, shallow} from 'enzyme';
import React from 'react';
import theme from './TestTheme';

export const getRender = (Component) => (props = {}, context = {theme}) => shallow(<Component {...props} />, {context});
export const getMount = (Component) => (props = {}, context = {theme}) => mount(<Component {...props} />, {context});
export const snapshot = (renderedComponent) => toJson(renderedComponent);
