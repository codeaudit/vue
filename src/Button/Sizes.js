export const small = 0.75;
export const normal = 1;
export const large = 1.5;
