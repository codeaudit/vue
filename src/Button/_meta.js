import examples from './examples';

export default {
    name: 'Button',
    keywords: ['basic', 'important', 'alt', 'special'],
    'see also': [],
    readme: require('./README.md'),
    status: 'experimental',
    componentsSources: [
        {
            name: 'Button',
            code: require('!raw!./Button')
        }
    ],
    examples
};
