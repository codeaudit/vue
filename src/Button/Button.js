import React, {Component, PropTypes} from 'react';
import Radium from './../utilities/Radium';
import {create} from './../styles/Transitions';
import {darken, getForegroundForBackground} from './../utilities/colorManipulator';
import * as ButtonSizes from './Sizes';
import * as ButtonTypes from './Types';

const darkenInvert = (foreground, background) => {
    const inverseBasicColorMultiplier = 0.35;
    const darkenedColor = darken(background, inverseBasicColorMultiplier);
    return {
        ':hover': {
            background: darkenedColor,
            border: `1px solid ${darkenedColor}`,
            color: foreground
        },
        background,
        border: `1px solid ${background}`,
        color: foreground
    };
};

class Button extends Component {
    static propTypes = {
        size: PropTypes.number,
        text: PropTypes.string,
        type: PropTypes.oneOf(ButtonTypes),
        onClick: PropTypes.func
    };
    static defaultProps = {
        size: ButtonSizes.normal,
        text: '',
        type: ButtonTypes.standard,
        onClick: () => {
        }
    };
    static contextTypes = {
        theme: PropTypes.shape({
            borders: PropTypes.shape({normalRadius: PropTypes.number.isRequired}),
            color: PropTypes.shape({
                alt: PropTypes.string,
                basic: PropTypes.string,
                darkInverse: PropTypes.string,
                important: PropTypes.string,
                lightInverse: PropTypes.string,
                normalBackground: PropTypes.string,
                textPrimary: PropTypes.string,
                transparent: PropTypes.string
            }),
            typography: PropTypes.shape({
                basicFontFamily: PropTypes.string,
                bold: PropTypes.number,
                lineHeightLarge: PropTypes.number.isRequired,
                small: PropTypes.number.isRequired
            }).isRequired
        }).isRequired
    };

    constructor(...rest) {
        super(...rest);
        this.getStyles = this.getStyles.bind(this);
        this.getStylesBasedOnType = this.getStylesBasedOnType.bind(this);
    }

    getStyles() {
        const {
            border: {normalRadius},
            color: {transparent},
            typography: {
                basicFontFamily,
                bold,
                lineHeightLarge,
                small
            }
        } = this.context.theme;
        const {size} = this.props;

        const fontSize = small * size;
        const height = Math.ceil(fontSize * lineHeightLarge);
        const borderRadiusMultiplier = 2;
        const borderRadius = normalRadius * borderRadiusMultiplier;
        const typeStyles = this.getStylesBasedOnType();

        return {
            root: {
                ':focus': {outline: 'none'},
                alignItems: 'flex-start',
                border: `1px solid ${transparent}`,
                borderRadius: `${borderRadius}px`,
                boxSizing: 'border-box',
                cursor: 'pointer',
                display: 'inline-block',
                fontFamily: basicFontFamily,
                fontSize: `${fontSize}px`,
                fontWeight: bold,
                height: `${height}px`,
                letterSpacing: '0.03em',
                lineHeight: `${lineHeightLarge}px`,
                margin: '0',
                padding: '0 1em',
                textAlign: 'center',
                textShadow: 'none',
                transition: create('0.5s'),
                whiteSpace: 'no-wrap',
                ...typeStyles
            }
        };
    }

    getStylesBasedOnType() {
        const {
            color: {
                alt,
                basic,
                important,
                darkInverse,
                lightInverse,
                normalBackground,
                textPrimary
            }
        } = this.context.theme;
        const {type} = this.props;
        const inverseColors = [darkInverse, lightInverse];

        if (type === ButtonTypes.basic) {
            return darkenInvert(darkInverse, basic);
        }
        if (type === ButtonTypes.important) {
            return darkenInvert(darkInverse, important);
        }
        if (type === ButtonTypes.alt) {
            return darkenInvert(darkInverse, alt);
        }
        if (type === ButtonTypes.basicAlt) {
            return {
                ':hover': {
                    background: basic,
                    border: `1px solid ${basic}`,
                    color: darkInverse
                },
                background: normalBackground,
                border: `1px solid ${textPrimary}`,
                color: textPrimary
            };
        }

        const inverseBackground = getForegroundForBackground(normalBackground, inverseColors);
        const inverseForeground = getForegroundForBackground(inverseBackground, inverseColors);
        return {
            ':hover': {
                background: inverseBackground,
                color: inverseForeground
            },
            background: normalBackground,
            border: `1px solid ${textPrimary}`,
            color: textPrimary
        };
    }

    render() {
        const {
            onClick,
            text
        } = this.props;
        const styles = this.getStyles();
        return (
            <button
                style={styles.root}
                onClick={onClick}
            >
                {text}</button>
        );
    }
}
export default Radium(Button);
