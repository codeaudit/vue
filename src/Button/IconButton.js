import React, {PropTypes} from 'react';
import ui from 'redux-ui';
import {normal} from './Sizes';
import Radium from './../utilities/Radium';
import transparent from './../utilities/Transparent';
import {create as createTransition} from '../utilities/Transitions';
import {createConditionalEventHandler, createEventHandlerIgnoringEventData} from './../utilities/component';

const getStyles = (props, theme) => ({
    root: {
        backgroundColor: (Boolean(props.disabled) || (!props.hovered && !props.ui.hovered)) ? props.backgroundColor : props.hoverBackgroundColor,
        border: (props.disabled && Boolean(props.border)) ? `1px solid ${theme.disabledPrimaryColor}` : Boolean(props.border) ? props.border : `1px solid ${transparent}`,
        borderRadius: props.circle ? '50%' : '0px',
        cursor: props.disabled ? 'not-allowed' : 'pointer',
        display: 'inline-block',
        lineHeight: 0.6,
        transition: props.transition,
    },
});

const IconButton = (props, context) => {
    const {
        color,
        disabled,
        hoverColor,
        hovered,
        ui,
        onClick,
        size,
        updateUI,
    } = props;
    const {
        baseIconSize,
        disabledPrimaryColor,
        smallGutter,
    } = context.theme;

    const width = baseIconSize * size;
    const iconColor = disabled ? disabledPrimaryColor : color;
    const iconHoverColor = disabled ? disabledPrimaryColor : hoverColor;
    const handleClick = createConditionalEventHandler(!disabled)(onClick);
    const handleMouseEnter = createEventHandlerIgnoringEventData(updateUI, 'hovered', true);
    const handleMouseLeave = createEventHandlerIgnoringEventData(updateUI, 'hovered', false);
    const styles = getStyles(props, context.theme);

    return (
        <div
            style={styles.root}
            onClick={handleClick}
            onMouseEnter={handleMouseEnter}
            onMouseLeave={handleMouseLeave}
        >
            {React.createElement(props.icon, {
                color: iconColor,
                hoverColor: iconHoverColor,
                hovered: hovered || ui.hovered,
                padding: smallGutter,
                width,
            })}
        </div>
    );
};
IconButton.propTypes = {
    /**
     * Background color of button
     */
    backgroundColor: PropTypes.string,
    /**
     * Border of icon button
     */
    border: PropTypes.string,
    /**
     * Will render a circular button when true
     */
    circle: PropTypes.bool,
    /**
     * Color of the SvgIcon within button
     */
    color: PropTypes.string,
    /**
     * Disables the button from responding to event handlers
     */
    disabled: PropTypes.bool,
    /**
     * Background color when hovered
     */
    hoverBackgroundColor: PropTypes.string,
    /**
     * Hover color of SvgIcon
     */
    hoverColor: PropTypes.string,
    /**
     * Indicates the IconButton is in a hovered state when true
     */
    hovered: PropTypes.bool,
    /**
     * Icon to render within button
     */
    icon: PropTypes.func.isRequired,
    /**
     * Click event handler; fired once a button is clicked
     */
    onClick: PropTypes.func,
    /**
     * Numeric value used as a multiplier to the button's size; 0.75, 1, and 1.5 as examples
     */
    size: PropTypes.number,
    transition: PropTypes.string,
    /**
     * Managed UI state props; can be overridden
     */
    ui: PropTypes.object,
};
IconButton.defaultProps = {
    backgroundColor: transparent,
    circle: false,
    disabled: false,
    onClick: () => {
    },
    size: normal,
    transition: createTransition('0.25s', 'all', '0ms', 'linear'),
};
IconButton.contextTypes = {
    theme: PropTypes.object.isRequired,
};
IconButton.displayName = 'IconButton';
export default Radium(ui({
    state: {
        hovered: false,
    }
})(IconButton));
