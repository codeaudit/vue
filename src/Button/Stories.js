import React from 'react';
import {storiesOf} from '@kadira/storybook';
import Button from './';
import * as ButtonSizes from './Sizes';
import * as ButtonTypes from './Types';

storiesOf('Button')
    .addWithInfo('empty/blank',
        `Button with no text or content`,
        () => (
            <Button />
        )
    )
    .addWithInfo('standard button',
        `Buttons default to a standard type`,
        () => (
            <Button text="Click Me" type={ButtonTypes.standard} />
        )
    )
    .addWithInfo('small, normal, large standard buttons',
        `size is a numeric value that will scale the button by times the size's specified amount`,
        () => (
            <div>
                <Button text="Click Me" size={ButtonSizes.small} />
                <Button text="Click Me" type={ButtonSizes.normal} />
                <Button text="Click Me" type={ButtonSizes.large} />
            </div>
        )
    )
    .addWithInfo('basic button type',
        ``,
        () => (
            <Button text="Click Me" type={ButtonTypes.basic} />
        )
    );
