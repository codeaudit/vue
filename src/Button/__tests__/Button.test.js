import Button from '../Button';
import {getRender, snapshot} from './../../../specHelpers/rendering';
import * as ButtonTypes from '../Types';

const renderButton = getRender(Button);
const clickEvent = {test: true};

test('button is click-able with an event handler', () => {
    const onClick = jest.fn();
    const button = renderButton({
        onClick,
    });
    simulateClick(button, clickEvent);
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith(clickEvent);
});

test('standard button', () => {
    const buttonWithDefaultType = renderButton({
        text: 'Click me',
    });
    expect(snapshot(buttonWithDefaultType)).toMatchSnapshot();

    const button = renderButton({
        text: 'Click me',
        type: ButtonTypes.standard,
    });
    expect(snapshot(button)).toMatchSnapshot();
});

test('basic button', () => {
    const button = renderButton({
        text: 'Click me',
        type: ButtonTypes.basic
    });
    expect(snapshot(button)).toMatchSnapshot();
});

test('important button', () => {
    const button = renderButton({
        text: 'Click me',
        type: ButtonTypes.important,
    });
    expect(snapshot(button)).toMatchSnapshot();
});

test('alternative button', () => {
    const button = renderButton({
        text: 'Click me',
        type: ButtonTypes.alt,
    });
    expect(snapshot(button)).toMatchSnapshot();
});

test('basic alternative button', () => {
    const button = renderButton({
        text: 'Click me',
        type: ButtonTypes.basicAlt,
    });
    expect(snapshot(button)).toMatchSnapshot();
});

test('special button', () => {
    const button = renderButton({
        text: 'Click me',
        type: ButtonTypes.special,
    });
    expect(snapshot(button)).toMatchSnapshot();
});

test('button can be resized', () => {
    const button = renderButton({size: 0.75});
    expect(snapshot(button)).toMatchSnapshot();
});

test('button can be disabled', () => {
    let onClick = jest.fn();
    const button = renderButton({
        disabled: true,
        onClick
    });
    simulateClick(button);
    expect(onClick).not.toBeCalled();
    expect(snapshot(button)).toMatchSnapshot();
});

function simulateClick(wrapper, evt = {}) {
    wrapper.simulate('click', evt);
}
