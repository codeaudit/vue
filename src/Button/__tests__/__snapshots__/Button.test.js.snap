exports[`test alternative button 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#eaab00",
      "border": "1px solid #eaab00",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#ffffff",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  }>
  Click me
</button>
`;

exports[`test basic alternative button 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#ffffff",
      "border": "1px solid #31363e",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#31363e",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  }>
  Click me
</button>
`;

exports[`test basic button 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#00a9e0",
      "border": "1px solid #00a9e0",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#ffffff",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  }>
  Click me
</button>
`;

exports[`test button can be disabled 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#ffffff",
      "border": "1px solid #31363e",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#31363e",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  } />
`;

exports[`test button can be resized 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#ffffff",
      "border": "1px solid #31363e",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#31363e",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "10.5px",
      "fontWeight": undefined,
      "height": "24px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  } />
`;

exports[`test important button 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#ea6c02",
      "border": "1px solid #ea6c02",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#ffffff",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  }>
  Click me
</button>
`;

exports[`test special button 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#31363e",
      "border": "1px solid #31363e",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#ffffff",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  }>
  Click me
</button>
`;

exports[`test standard button 1`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#ffffff",
      "border": "1px solid #31363e",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#31363e",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  }>
  Click me
</button>
`;

exports[`test standard button 2`] = `
<button
  data-radium={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "background": "#ffffff",
      "border": "1px solid #31363e",
      "borderRadius": "6px",
      "boxSizing": "border-box",
      "color": "#31363e",
      "cursor": "pointer",
      "display": "inline-block",
      "fontFamily": "\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif",
      "fontSize": "14px",
      "fontWeight": undefined,
      "height": "32px",
      "letterSpacing": "0.03em",
      "lineHeight": "2.285",
      "margin": "0",
      "padding": "0 1em",
      "textAlign": "center",
      "textShadow": "none",
      "transition": "all 0.5s linear 0ms",
      "whiteSpace": "no-wrap",
    }
  }>
  Click me
</button>
`;
